File: kotlin/Client/src/main/kotlin/com/github/csandiego/pocopenapitools/client/openapitools/OpenApiToolsBookService.kt
package com.github.csandiego.pocopenapitools.client.openapitools

import com.github.csandiego.pocopenapitools.client.data.Book
import com.github.csandiego.pocopenapitools.client.service.BookService
import com.github.csandiego.pocopenapitools.openapitools.models.Book as DTO

class OpenApiToolsBookService(private val api: OpenApiToolsBookApi) : BookService {

    override fun list(): List<Book> {
	return api.list().map { Book(it.id, it.title, it.author) }
    }

    override fun create(book: Book) {
	api.create(DTO(book.title, book.author, book.id))
    }

    override fun get(id: Int): Book {
	return api.get(id).run { Book(id, title, author) }
    }

    override fun update(id: Int, book: Book) {
	api.update(id, DTO(book.title, book.author, book.id))
    }

    override fun delete(id: Int) {
	api.delete(id)
    }
}

File: swift/Client/Sources/Lib/OpenApiToolsBookService.swift
import OpenApiTools

public class OpenApiToolsBookService: BookService {

    let api: OpenApiToolsBookAPI.Type

    public init(api: OpenApiToolsBookAPI.Type) {
        self.api = api
    }

    public func list(completion: @escaping ((_ books: [Book]?, _ error: Error?) -> Void)) {
        api.list { dto, error in
            completion(dto?.map { Book(id: $0.id, title: $0.title, author: $0.author) }, error)
        }
    }

    public func create(book: Book, completion: @escaping ((_ data: Void?, _ error: Error?) -> Void)) {
        api.create(book: OpenApiTools.Book(id: book.id, title: book.title, author: book.author)) { data, error in
            completion(data, error)
        }
    }

    public func callGet(id: Int, completion: @escaping ((_ book: Book?, _ error: Error?) -> Void)) {
        api.callGet(id: id) { dto, error in
            var book: Book? = nil
            if let dto = dto {
                book = Book(id: dto.id, title: dto.title, author: dto.author)
            }
            completion(book, error)
        }
    }

    public func update(id: Int, book: Book, completion: @escaping ((_ data: Void?, _ error: Error?) -> Void)) {
        api.update(id: id, book: OpenApiTools.Book(id: book.id, title: book.title, author: book.author)) { data, error in
            completion(data, error)
        }
    }

    public func delete(id: Int, completion: @escaping ((_ data: Void?, _ error: Error?) -> Void)) {
        api.delete(id: id) { data, error in
            completion(data, error)
        }
    }
    
}
